### PR TITLE
Issue 28: add remote state

### PR DESCRIPTION
**Pull request by rick-c-goog**
Add remote state to the 2 clusters:
1. make create CLUSTER
2. make secure CLUSSTER

need to add the arguments in Makefile, also changes in 
scripts/create_cluster.sh
scripts/secure_cluster.sh

The error at the end GET_CREDS="$(terraform output --state=./terraform/$1/terraform.tfstate get_credentials)
that can be ignored for now.

### CODE REVIEW

**Review by knee-berts — COMMENTED** [review]
Looks great so far. I think we should append a random ID to the storage bucket name to ensure that the storage bucket name is always globally unique.

**Comment by rick-c-goog** [comment]
@googlebot I signed it!

**Comment by rick-c-goog** [comment]
Close it for now, will do a new one later on.

File: scripts/destroy_cluster.sh [hunk 1]
```diff
@@ -22,14 +22,25 @@ ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )/.." && pwd )"
...
 
-# Remove the tfvars file generated during "make create"
-rm -f "$ROOT/terraform/cluster_build/terraform.tfvars"
-rm -f "$ROOT/terraform/cluster_build/terraform.tfstate"
+if [ "$STATE" == gcs ]; then 
+   cd "${ROOT}/terraform/cluster_build" 
+   sed -i "s/local/gcs/g" backend.tf
+   (cd "${ROOT}/terraform/cluster_build"; terraform init -input=false -backend-config="bucket=$BUCKET")
+   terraform state rm "module.kms" 
+   (cd "${ROOT}/terraform/cluster_build"; terraform destroy -input=false -auto-approve)
```
Comment (jennviau) [inline scripts/destroy_cluster.sh:36]: I think we still need to at least remove the tfvars file during the destroy, ideally we're going to remove the state too.
Also wondering about what happens to the bucket?

File: scripts/set-env.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export PROJECT=$(gcloud config get-value project)
+export ZONE=us-central1-a
```
Comment (jennviau) [inline scripts/set-env.sh:2]: Might consider changing zone to also use the gcloud command
`gcloud config get-value compute/zone`

File: scripts/destroy_cluster.sh (hunk 1 above)
Comment (rick-c-goog) [reply scripts/destroy_cluster.sh]: Add the following
rm -f "$ROOT/terraform/cluster_build/terraform.tfvars"   
gsutil -m rm -r gs://$BUCKET

File: scripts/set-env.sh (hunk 2 above)
Comment (rick-c-goog) [reply scripts/set-env.sh]: updated,

**Comment by knee-berts** [comment]
Please run the `terraform fmt` command against the **/terraform/cluster_build** directory and then resubmit the Pull Request. It is required that `terraform fmt` be executed against all Terraform configuration files in order to provide consistency across the repo.

**Comment by rick-c-goog** [comment]
terraform fmt ran, commit pushed to forked main branch,

**Comment by knee-berts** [comment]
Please run the `terraform fmt` command against the **/terraform/security** directory and then resubmit the Pull Request. It is required that `terraform fmt` be executed against all Terraform configuration files in order to provide consistency across the repo.

**Comment by knee-berts** [comment]
An error was detected executing `terraform init` against the **/terraform/cluster_build** directory. Please run `terraform init` locally against that directory to resolve any errors and then resubmit the Pull Request.

**Comment by knee-berts** [comment]
Prechecks Complete - Ready for review

File: scripts/set-env.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export PROJECT=$(gcloud config get-value project)
+export ZONE=$(gcloud config get-value compute/zone)
+export GOVERNANCE_PROJECT=$(gcloud config get-value project)
+export WINDOWS_CLUSTER=false
+export PUBLIC_CLUSTER=false
+export STATE=gcs
```
Comment (jennviau) [inline scripts/set-env.sh:6]: I would recommend defaulting to local state. Can you add a small section to the appropriate readme ( for now) on how to leverage the Remote State functionality?

File: scripts/generate-cluster-tfvars.sh [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Copyright 2020 Google LLC
+# Copyright 2019 Google LLC
```
Comment (jennviau) [inline scripts/generate-cluster-tfvars.sh:2]: Can you update the date on the license section, it reverted from 2020 to 2019

File: scripts/create_cluster.sh [hunk 5]
```diff
@@ -22,14 +22,32 @@ ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )/.." && pwd )"
...
 
 # Initialize and run Terraform
-(cd "${ROOT}/terraform/cluster_build"; terraform init -input=false)
-(cd "${ROOT}/terraform/cluster_build"; terraform apply -input=false -auto-approve)
-
-# Get cluster credentials
-GET_CREDS="$(terraform output --state=./terraform/$1/terraform.tfstate get_credentials)"
+if [ "$STATE" == gcs ]; then
+  cd "${ROOT}/terraform/cluster_build"
+  BUCKET=$PROJECT-$CLUSTER_TYPE-cluster-tf-state
```
Comment (jennviau) [inline scripts/create_cluster.sh:32]: This is probably only a small likelyhood of impacting  but GCS buckets have a 63 character max name. might consider just $PROJECT-$CLUSTER_TYPE-state

**Review by jennviau — COMMENTED** [review]
@rick-c-goog  This looks great, I left a few small comments. The only one that I think is super necessary are around the license date, and defaulting to local state vs defaulting to remote state

File: scripts/create_cluster.sh (hunk 5 above)
Comment (rick-c-goog) [reply scripts/create_cluster.sh]: updated

File: scripts/set-env.sh (hunk 3 above)
Comment (rick-c-goog) [reply scripts/set-env.sh]: updated

**Comment by knee-berts** [comment]
Prechecks Complete - Ready for review

**Comment by jennviau** [comment]
Thanks @rick-c-goog 
lgtm

**Review by jennviau — APPROVED** [review]

**Pull request merged**